File: QuickSort.py
def quickSort(indice):
    quickSortTam(indice, len(indice)-1)
  
def quickSortTam(indice, primeiro, ultimo):
	if primeiro < ultimo:
		
		pivo = particao(indice, primeiro, ultimo)
     
		quickSortTam(indice, primeiro, pivo - 1)
    	quickSortTam(indice, pivo+1, ultimo)
		
def particao(indice, primeiro, ultimo):
	pivoNum = indice[primeiro]
  
	esquerda = primeiro + 1
	direita = ultimo
  
  	flag = False
 	while not flag:
    
    	while esquerda <= direita and indice[esquerda] <= pivoNum:
			esquerda = esquerda + 1
		
		while indice[direita] >=pivoNum and direita >= esquerda:
			direita = direita - 1
	
		if direita < esquerda:
			flag = True
		else:
			swap = indice[esquerda]
			indice[esquerda] = indice[direita]
			indice[direita] = swap
			
	swap = indice[primeiro]
	indice[primeiro] = indice[direita]
	indice[direita] = swap

	return direita

indice = [54, 26, 93, 17, 77, 31, 44, 55, 20, 23]

# implementar tempo de execução para teste de eficiência

quickSort(indice)
print(indice)

File: QuickSort.rb
def quick(vetor, inicio, fim)
  if inicio < fim
    pivo = particao(vetor, inicio, fim)
    quick(vetor, inicio, pivo -1)
    quick(vetor, pivo + 1, fim)
  end
  vetor
end

def particao(vetor, inicio, fim)
  vetorAtual = inicio
  i = inicio
  while i < fim do
    if vetor[i] <= vetor[fim]
      swap(vetor, i, vetorAtual)
      vetorAtual += 1
    end
    i += 1
  end
  swap(vetor, fim, vetorAtual)
  vetorAtual
end

def swap(vetor, primeiro, ultimo)
  temp = vetor[primeiro]
  vetor[primeiro] = vetor[ultimo]
  vetor[ultimo] = temp
end

puts quick([54, 26, 93, 17, 77, 31, 44, 55, 20, 23],0,9).inspect 
